packages/automerge-repo/src/synchronizer/DocSynchronizer.ts

```diff
@@ -1,178 +1,156 @@
+    this.#log(`sendSyncMessage ->${peerId}`)
 
-      conciseLog(`${peerId} ⬅️️  ${this.handle.documentId} (${message.byteLength} bytes)`)
-      if (opsLog.enabled) { // guard opsLog, so decodeChange is not called unnecessarily, because it can be expensive
-        opsLog(decoded.changes.flatMap(change => {
-          return decodeChange(change).ops.map((op) => JSON.stringify(op))
-        }))
-      }
+    // TODO: Right now this can send duplicate messages even if nothing has changed
+
+    const syncState = this.#getSyncState(peerId)
+    const [newSyncState, message] = A.generateSyncMessage(doc, syncState)
+    this.#setSyncState(peerId, newSyncState)
+    if (message) {
+      this.logMessage(`sendSyncMessage 🡒 ${peerId}`, message)
 
-      const channelId = this.handle.documentId as unknown as ChannelId
+      const channelId = this.#handle.documentId as string as ChannelId
       this.emit("message", {
         targetId: peerId,
         channelId,
         message,
         broadcast: false,
       })
     } else {
-      log(
-        `[${this.handle.documentId}]->[${peerId}]: sendSyncMessage: [no message generated]`
-      )
+      this.#log(`sendSyncMessage ->${peerId} [no message generated]`)
     }
   }
 
+  logMessage = (label: string, message: Uint8Array) => {
+    const size = message.byteLength
+    const logText = `${label} ${size}b`
+    const decoded = A.decodeSyncMessage(message)
+
+    this.#conciseLog(logText)
+    this.#log(logText, decoded)
+
+    // expanding is expensive, so only do it if we're logging at this level
+    const expanded = this.#opsLog.enabled
+      ? decoded.changes.flatMap(change =>
+          A.decodeChange(change).ops.map(op => JSON.stringify(op))
+        )
+      : null
+    this.#opsLog(logText, expanded)
+  }
+
+  /// PUBLIC
+
+  hasPeer(peerId: PeerId) {
+    return this.#peers.includes(peerId)
+  }
+
   async beginSync(peerId: PeerId) {
-    log(`[${this.handle.documentId}]: beginSync: ${peerId}`)
-    const { documentId } = this.handle
-    const doc = await this.handle.syncValue()
-
-    // Just in case we have a sync state already, we round-trip it through
-    // the encoding system to make sure state is preserved. This prevents an
-    // infinite loop caused by failed attempts to send messages during disconnection.
-    // TODO: we should be storing sync states and besides, we only need to do this on reconnect
-    this.setSyncState(
-      peerId,
-      Automerge.decodeSyncState(
-        Automerge.encodeSyncState(this.getSyncState(peerId))
-      )
-    )
+    this.#log(`beginSync: ${peerId}`)
+    const doc = await this.#handle.provisionalValue()
 
-    this.sendSyncMessage(peerId, documentId, doc)
+    // HACK: if we have a sync state already, we round-trip it through the encoding system to make
+    // sure state is preserved. This prevents an infinite loop caused by failed attempts to send
+    // messages during disconnection.
+    const syncStateRaw = this.#getSyncState(peerId)
```

Comment:
we should save the sync states but only for peers which say they're likely to come back: ephemeral tabs in the browser for example shouldn't have their sync states saved because they can't reconnect, if they reload they come back empty with a new peerID

Reply:
note that for ephemeral tabs we should support a read-only mode for storage. this would allow ephemeral processes to load off "disk" and then present themselves for sync already caught up to whatever the local state is.